pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,342 @@
+}
+
+func writeReceivedProtoMessageLocalProfile(localProfile *profile.LocalProfile) {
+	contents, err := proto.Marshal(localProfile)
+	if err != nil {
+		log.Errorf("writeReceivedProtoMessageLocalProfile Marshalling failed: %s", err)
+		return
+	}
+	writeProtoMessage(savedLocalProfileFile, contents)
+	return
+}
+
+//parseProfile process local and global profile configuration
+//must be called before processing of app instances from config
+func parseProfile(ctx *getconfigContext, config *zconfig.EdgeDevConfig) {
+	log.Functionf("parseProfile start: globalProfile: %s localProfile: %s",
+		ctx.globalProfile, ctx.localProfile)
+	if ctx.globalProfile != config.GlobalProfile {
+		log.Noticef("parseProfile: GlobalProfile changed from %s to %s",
+			ctx.globalProfile, config.GlobalProfile)
+		ctx.globalProfile = config.GlobalProfile
+	}
+	if ctx.localProfileServer != config.LocalProfileServer {
+		log.Noticef("parseProfile: LocalProfileServer changed from %s to %s",
+			ctx.localProfileServer, config.LocalProfileServer)
+		ctx.localProfileServer = config.LocalProfileServer
+		processLocalProfileServerChange(ctx)
+	}
+	ctx.profileServerToken = config.ProfileServerToken
+	checkAndHandleCurrentProfileUpdate(ctx)
+	log.Functionf("parseProfile done globalProfile: %s currentProfile: %s",
+		ctx.globalProfile, ctx.currentProfile)
+}
+
+//determineCurrentProfile return current profile based on localProfile, globalProfile
+func determineCurrentProfile(ctx *getconfigContext) string {
+	if ctx.localProfile == "" {
+		return ctx.globalProfile
+	}
+	return ctx.localProfile
+}
+
+//checkAndHandleCurrentProfileUpdate calculates and sets current profile and publishes ZedAgentStatus on change
+func checkAndHandleCurrentProfileUpdate(ctx *getconfigContext) {
+	currentProfile := determineCurrentProfile(ctx)
+	if ctx.currentProfile != currentProfile {
+		log.Noticef("checkAndHandleCurrentProfileUpdate: profile changed from %s to %s",
+			ctx.currentProfile, currentProfile)
+		ctx.currentProfile = currentProfile
+		publishZedAgentStatus(ctx)
+	}
+}
+
+//triggerGetLocalProfile notifies task to reload local profile from profileServer
+func triggerGetLocalProfile(ctx *getconfigContext) {
+	log.Functionf("triggerGetLocalProfile")
+	select {
+	case ctx.localProfileTrigger <- Notify{}:
+	default:
+	}
+}
+
+//processLocalProfileServerChange on first call runs readSavedLocalProfile if localProfileServer defined
+//on the next calls or if localProfileServer is not defined it runs cleanSavedProtoMessage, clean localProfile
+//and sends notification to reload localProfile from localProfileServer
+func processLocalProfileServerChange(ctx *getconfigContext) {
+	if !ctx.callProcessLocalProfileServerChange {
+		ctx.callProcessLocalProfileServerChange = true
+		if ctx.localProfileServer != "" {
+			// read saved config on the first call
+			localProfile, err := readSavedLocalProfile(ctx, false)
+			if err != nil {
```

Comment:
Why can't we readsavelocalprofile at the top of the task? That is a lot simpler.

Reply:
It called from the place where we have all needed information and after which we need to know the local profile from saved if exists.

Reply:
We don't need that information to read the file and set localProfile to what was read. When we parse we'll get a globalProfile and localProfileServer and we can use that to determine the initial currentProfile.